### PR TITLE
Fix compilation of tests with `/std:c++17`

### LINKED ISSUE #2089: Pybind fails to build with MSVC 19.16 in C++17 mode
## Fails to build with MSVC 19.16.27034.0 in C++17
I used pybind v 2.4.3

> cmake -G Ninja -DPYBIND11_CPP_STANDARD=/std:c++17 ..\src
> ninja

Error:
> D:\Libs\Pybind\2.4.3\src\include\pybind11/stl_bind.h(415): error C2893: Failed to specialize function template 'enable_if<_Test,_Ty>::type pybind11::detail::vector_buffer(Class_ &)'
        with
        [
            _Ty=void
        ]

No problem with the default c++14 mode.

Great library, BTW! 😃 👍

### PR DESCRIPTION
**Pull request by YannickJadoul**
Seems it needs to be closed to get it passed correctly?

Closes #2089

So this first commit should fail, given #2089.

### CODE REVIEW

**Comment by YannickJadoul** [comment]
Am I going crazy? The debug runs print `PYBIND11_CPP_STANDARD=/std:c++17` from CMake, but the actual compile command contains `/std:c++14`???

**Comment by YannickJadoul** [comment]
Seriously!?

https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html

> $<COMPILE_LANGUAGE:languages>
> ...
> Note that with Visual Studio Generators and Xcode there is no way to represent target-wide compile definitions or include directories separately for C and CXX languages. Also, with Visual Studio Generators there is no way to represent target-wide flags separately for C and CXX languages. Under these generators, expressions for both C and C++ sources will be evaluated using CXX if there are any C++ sources and otherwise using C. A workaround is to create separate libraries for each source file language instead:

**Comment by henryiii** [comment]
The correct solution probably would be to use compile_features; when we make this 3.7+ only we should be able to change this properly.

**Comment by YannickJadoul** [comment]
> The correct solution probably would be to use compile_features; when we make this 3.7+ only we should be able to change this properly.

Yes, I had thought of that as well. But depends on dropping old CMake support, then.

> But yes, this is a general issue that always needs a workaround for these IDE's.

Good point. I'll check if there's more of these around!

(Should I also check for Xcode, btw, or not?)

**Comment by henryiii** [comment]
If it doesn't work on Xcode, I'd likely know pretty quickly. ;)

I can't think of any more of these that are likely to be a problem, but I wouldn't have remembered the one above (I had to fix in in CLI11 a long time ago). I know of some that probably don't affect us (like the `...` syntax in version support being broken in MSVC's forked copy of CMake 3.11)

**Comment by YannickJadoul** [comment]
Thanks! I found 2 more occurences of `$<COMPILE_LANGUAGE:CXX>`: one for the embedding target (same fix), and one here: https://github.com/pybind/pybind11/blob/8b9eb964d904c1dab33ed819497fb997f25e473b/tools/pybind11Tools.cmake#L257

Not sure what the last one does, but it's already MSVC-specific. Any idea what to do? Just ignore, since it's only about parallel builds?

**Comment by henryiii** [comment]
Maybe tag it with a TODO? Is there a reason that `cmake --build -j<N>` doesn't work for parallel builds? Manually sticking specific compile options is usually a bad idea. And, due to the COMPILE_LANGUAGE, this one should be a no-op.

**Comment by YannickJadoul** [comment]
> Maybe tag it with a TODO? Is there a reason that `cmake --build -j<N>` doesn't work for parallel builds? Manually sticking specific compile options is usually a bad idea. And, due to the COMPILE_LANGUAGE, this one should be a no-op.

No clue. Never really touched the CMake internals before. I'll just apply the same fix anyway. If someone passes by that knows more about CMake, we can still change it? :-)

**Comment by YannickJadoul** [comment]
Builds are still using `/std:c+14`!? Am I somehow going crazy? I thought I had found the issue :-(

**Comment by bstaletic** [comment]
I know the title says `/std:c++17`, but could we maybe include this as well: https://github.com/pybind/pybind11/issues/2159

**Comment by YannickJadoul** [comment]
> I know the title says `/std:c++17`, but could we maybe include this as well: #2159

Yes, good point. I was planning to see if supporting VS 2019 would still be possible, but this is also something to add!
Once I get things to fail, that is ...

File: .appveyor.yml [hunk 1]
```diff
@@ -57,14 +56,15 @@ install:
     7z x 3.3.3.zip -y > $null
     $env:CMAKE_INCLUDE_PATH = "eigen-eigen-67e894c6cd8f;$env:CMAKE_INCLUDE_PATH"
 build_script:
+- set VERBOSE=1
+- cmake --version
 - cmake -G "%CMAKE_GENERATOR%" -A "%CMAKE_ARCH%"
-    -DPYBIND11_CPP_STANDARD=/std:c++%CPP%
+    -DPYBIND11_CPP_STANDARD=/std:c++%CPP
```
Comment (bstaletic) [inline .appveyor.yml:62]: You stripped the trailing `%` here. `CPP` isn't treated as a variable in this case.

Comment (YannickJadoul) [reply .appveyor.yml:62]: Thanks. Committed too quickly!

File: tools/pybind11Tools.cmake [hunk 2]
```diff
@@ -218,8 +220,10 @@ function(pybind11_add_module target_name)
   # Make sure C++11/14 are enabled
   if(PYBIND11_CPP_STANDARD)
     if(CMAKE_VERSION VERSION_LESS 3.3 OR CMAKE_GENERATOR MATCHES "Visual Studio")
+      message("HERE, I hope?")
       target_compile_options(${target_name} PUBLIC ${PYBIND11_CPP_STANDARD})
     else()
+      message("NOT HERE, I hope?")
```
Comment (LeslieGerman) [inline tools/pybind11Tools.cmake:226]: Have you left these messages in by intent ? (on line 223, too)

**Comment by LeslieGerman** [comment]
@YannickJadoul 
> Thanks! I found 2 more occurences of `$<COMPILE_LANGUAGE:CXX>`: one for the embedding target (sa
> target_compile_options(${target_name} PRIVATE $<$<NOT:$<CONFIG:Debug>>:$<$<COMPILE_LANGUAGE:CXX>:/MP>>) 
> Not sure what the last one does ...

[/MP: Build with Multiple Processes](https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2017)

According to my experiences in previous years, it may or may not improve compilation time. See also [MS own Guideline](https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2017#guidelines).

If compilation speed is not a key factor here, you can safely omit this option IMO.

File: tools/pybind11Tools.cmake (hunk 2 above)
Comment (YannickJadoul) [reply tools/pybind11Tools.cmake:226]: Yes. This is in the commit with message "Debugging non-failure of /std:c++17", and this PR is not ready, yet.

Comment (YannickJadoul) [reply tools/pybind11Tools.cmake:226]: Yes. This is in the commit with message "Debugging non-failure of /std:c++17", and this PR is not ready, yet.

**Comment by YannickJadoul** [comment]
> [/MP: Build with Multiple Processes](https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2017)

Yes, that I can look up. My question was just wondering out loud why it's in pybind11's CMake and why not in debug, etc. (Which I might be able to look up through some `git blame`, yes.)

**Comment by bstaletic** [comment]
What's the status of this? Are we just going to mandate `/permissive-`?

**Comment by YannickJadoul** [comment]
Yes, but I haven't gotten to cleaning up this PR and adding it to the docs. Feel free to take over, if you want?

**Pull request closed without merging**